contrib/server/operations/pbench-upload-results.py

```diff
@@ -0,0 +1,270 @@
+    else:
+        when = ""
+
+    if parsed.tarball.is_dir():
+        what = f"DIRECTORY {parsed.tarball}"
+    else:
+        what = f"TARBALL {parsed.tarball}"
+
+    print(f"{what}{when} -> SERVER {parsed.server}")
+
+    checkpoint: Optional[Path] = None
+    processed: list[str] = []
+    if parsed.checkpoint:
+        checkpoint = parsed.checkpoint
+        if parsed.verify:
+            print(f"Processing checkpoint state from {checkpoint}...")
+        if checkpoint.exists():
+            processed = checkpoint.read_text().splitlines()
+            if parsed.verify:
+                print(f"[CPT] done {len(processed)}")
+        if parsed.verify:
+            print(
+                f"Finished processing checkpoint data: {len(processed)} checkpointed files"
+            )
+
+    if parsed.verify:
+        print("Identifying target tarballs")
+    skipped = 0
+    early = 0
+    late = 0
+    if parsed.tarball.is_dir():
+        pool = parsed.tarball.glob("**/*.tar.xz")
+        which = []
+        for t in pool:
+            if not t.is_file():
+                continue
+            date = t.stat().st_mtime
+            if since_ts and date < since_ts:
+                early += 1
+                continue
+            if before_ts and date >= before_ts:
+                late += 1
+                continue
+            if str(t) in processed:
+                skipped += 1
+                if parsed.verify:
+                    print(f"[CPT] skip {t}")
+                continue
+            which.append(t)
+    elif parsed.tarball.is_file():
+        which = [parsed.tarball]
+    else:
+        print(f"Path {parsed.tarball} doesn't exist", file=sys.stderr)
+        return 1
+    if parsed.verify:
+        print(
+            f"Identified {len(which)} target tarballs: skipped {skipped}, {early} too old, {late} too new"
+        )
+
+    checkwriter: Optional[TextIOWrapper] = None
+    if checkpoint:
+        checkwriter = checkpoint.open(mode="a")
+
+    if parsed.verify:
+        print("Uploading target files...")
+    success = 0
+    failure = 0
+    failures = set()
+    duplicate = 0
+    for t in which:
+        try:
+            if parsed.preview:
+                print(f"UPLOAD {t}")
+                success += 1
+                if checkwriter:
+                    print(t, file=checkwriter, flush=True)
+            else:
+                response = upload(parsed.server, parsed.token, t, metadata=metadata)
+                if response.ok:
+                    if response.status_code == HTTPStatus.OK:
+                        duplicate += 1
+                    else:
+                        success += 1
+                    if parsed.verify:
+                        print(f"UPLOAD {t}: {response.status_code}")
+                    if checkwriter:
+                        print(t, file=checkwriter, flush=True)
+                else:
+                    failure += 1
+                    failures.add(response.status_code)
```

Comment:
Alternatively, we could record the failures in a `dict[int, list[str]` with the status as the key and the value as a list of paths.  (This could be implemented straightforwardly using a [`defaultdict`](https://docs.python.org/3/library/collections.html#collections.defaultdict).)

Reply:
I didn't see any need to get fancy. Most likely we're only going to get connection failures or out-of-space errors anyway. (Unpleasantly, the first manifestation of "out of space" that I saw in my tests were from `nginx` throwing a big XML with a 500 because it can't buffer the data stream ...)